### PR TITLE
Adding dynamicString helper to help handle dynamic links for voter reg.

### LINKED ISSUE #657: Use the Card style for voter registration action.
### User Story
As a user, I want voter registration to look like an action so that I know I need to do it.

#### Additional Information
Pivotal ID: [#154525500](https://www.pivotaltracker.com/story/show/154525500)

image

### Why This Matters
In the world of multiple actions, we want to start thinking of Actions as visually distinct cards.

### PR DESCRIPTION
**Pull request by weerd**
### What does this PR do?
This PR updates the Voter Registration action. Since the designs call for a making this action a card, with content copy AND a button, I decided that it might be best to modify how we deal with dynamic content to generate the links. Thus, we now have the `dynamicString()` method available to help with replacing tokens in a string (like a URL for example) with specified info. This makes it a bit more flexible so we don't have to adhere to a specific structure in the string and it can now be utilized for _any_ strings, not just URLs.

It also removes the old `dynamicLink` approach, which seemed reasonable at the time given the constraints.

### What are the relevant tickets/cards?
Fixes #657

#### Small
screen shot 2018-02-06 at 4 38 17 pm

#### Medium
screen shot 2018-02-06 at 4 39 03 pm

#### Large
screen shot 2018-02-06 at 4 39 57 pm

### Checklist
- [x] Documentation added for new features/changed endpoints.
- [x] Added appropriate feature/unit tests.
- [x] Added screenshot to PR description of related front-end updates on **small** screens.
- [x] Added screenshot to PR description of related front-end updates on **medium** screens.
- [x] Added screenshot to PR description of related front-end updates on **large** screens.

### CODE REVIEW

File: resources/assets/components/VoterRegistration/VoterRegistration.js [hunk 1]
```diff
@@ -1,67 +1,68 @@
...
-        stepIndex={stepIndex}
+        step={stepIndex}
       />
       <FlexCell width="two-thirds">
-        <Markdown>{ formattedContent }</Markdown>
+        <Card className="rounded bordered voter-registration" title={title}>
+          <div className="padded clearfix">
+            <Markdown>{ content }</Markdown>
+
+            { link ? <a className="button" href={dynamicString(link, tokens)} target="blank">Start Registration</a> : null }
```
Comment (mendelB) [inline resources/assets/components/VoterRegistration/VoterRegistration.js]: Should `target="blank"` have the underscore before it? `target="_blank"`?

File: resources/assets/components/VoterRegistration/VoterRegistration.js [hunk 2]
```diff
@@ -1,67 +1,68 @@
...
     </Flex>
   );
 };
 
 VoterRegistration.propTypes = {
+  campaignId: PropTypes.string.isRequired,
+  campaignRunId: PropTypes.string.isRequired,
   content: PropTypes.string,
   hideStepNumber: PropTypes.bool,
+  link: PropTypes.string,
```
Comment (mendelB) [inline resources/assets/components/VoterRegistration/VoterRegistration.js]: Is there a specific reason `link` is optional? Do we envision a Voter reg action that just has the link within the `content` or something?

File: resources/assets/components/VoterRegistration/VoterRegistration.js (hunk 1 above)
Comment (mendelB) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: https://stackoverflow.com/a/35703051/5657278

Comment (weerd) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: haha, I was wondering too. I keep forgetting which one to use. I remember I asked this same question when you did one without the underscore. Also just found the [link](https://stackoverflow.com/questions/35703005/what-is-the-difference-between-target-blank-and-target-blank) you shared about it. I think you're right and _underscore_ is the way to go!

File: resources/assets/components/VoterRegistration/VoterRegistration.js (hunk 2 above)
Comment (weerd) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: I was debating that... initially I thought we'd want it to be required, but if we ever end up incorporating a widget or something for our own experience for voter registration, than I could see it being optional. 🤔  But up for debate... thoughts?

File: resources/assets/components/VoterRegistration/VoterRegistration.js (hunk 1 above)
Comment (mendelB) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: LOL! I remembered that we spoke about this before but I was too 🍻to find it

File: resources/assets/components/VoterRegistration/VoterRegistration.js (hunk 2 above)
Comment (DFurnes) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: If it's required for now (which it seems like it is), I'd say we mark it as required.

Comment (weerd) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: good call out! Chatted w/ @DFurnes too and consensus is to require it for now :)

Comment (mendelB) [reply resources/assets/components/VoterRegistration/VoterRegistration.js]: ugh sorry, I missed your initial response to this for some reason 😕

File: docs/content-publishing/actions/voter-registration-action.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+The `VoterRegistrationAction` component renders a visual component which allows DoSomething.org members to register to vote via a link to an external voter registration partner.
+
+![Voter Registration Action component](../_assets/voter-registration-action-component.png)
+
+
+## Usage Instructions
+When creating a new _Voter Registration Action_ on Contentful, you are required to enter a value for the `link` field.
+
+Enter a full URL where the member should be directed to upon clicking on the button in the start their registration.
```
Comment (mendelB) [inline docs/content-publishing/actions/voter-registration-action.md]: slight typo: `in the start their registration`

File: docs/content-publishing/actions/voter-registration-action.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+The resulting URL for the rendered button on the page will be:
+```http
+http://example.com?userId=551a2b3c4d5e6f7g8h9i0j1k2&campaign=1234&source=web
+```
+
+As another example, if you enter the following URL string into the `link` field:
+```http
+https://turbonotes.xyz/?r=user:{northstarId},campaign:{campaignId},source:{source}
```
Comment (mendelB) [inline docs/content-publishing/actions/voter-registration-action.md]: lol http://turbonote.com/

**Comment by lkpttn** [comment]
There's a visual issue with this card, the button text doesn't match up with the rest of the buttons we have. Should be a bold weight, but it's just normal.

**Comment by lkpttn** [comment]
Yeah it didn't make a comment in Slack like I wanted lol.

**Pull request merged**